File: src/main/java/xyz/chalky/taboo/commands/music/QueueSlashCommand.java
package xyz.chalky.taboo.commands.music;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.interactions.commands.build.SubcommandData;
import org.jetbrains.annotations.NotNull;
import xyz.chalky.taboo.central.Taboo;
import xyz.chalky.taboo.core.CommandFlag;
import xyz.chalky.taboo.core.SlashCommand;
import xyz.chalky.taboo.music.GuildAudioPlayer;

import java.time.Instant;
import java.util.List;
import java.util.concurrent.BlockingQueue;

public class QueueSlashCommand extends SlashCommand {

    public QueueSlashCommand() {
        setCommandData(
                Commands.slash("queue", "Queue a song")
                        .addSubcommands(
                                new SubcommandData("list", "List the current queue."),
                                new SubcommandData("clear", "Clear the current queue.")
                        )
        );
        addCommandFlags(CommandFlag.MUSIC);
        setEphemeral(false);
    }

    @Override
    public void executeCommand(@NotNull SlashCommandInteractionEvent event) {
        GuildAudioPlayer guildAudioPlayer = Taboo.getInstance().getAudioManager().getAudioPlayer(event.getGuild().getIdLong());
        BlockingQueue<AudioTrack> queue = guildAudioPlayer.getScheduler().getQueue();
        switch (event.getSubcommandName()) {
            case "list" -> {
                StringBuilder description = new StringBuilder("Queue:\n");
                int trackSize = queue.size();
                int trackCount = Math.min(trackSize, 10);
                List<AudioTrack> trackList = queue.stream().toList();
                for (int i = 0; i < trackCount; i++) {
                    AudioTrack track = trackList.get(i);
                    AudioTrackInfo info = track.getInfo();
                    description.append("`#")
                            .append(i + 1)
                            .append("` [")
                            .append(info.title)
                            .append("](")
                            .append(info.uri)
                            .append(") by ")
                            .append(info.author)
                            .append("\n");
                }
                if (trackSize > trackCount) {
                    description.append("And ")
                            .append("`")
                            .append(trackSize - trackCount)
                            .append("`")
                            .append(" more tracks...");
                }
                EmbedBuilder embed = new EmbedBuilder()
                        .setTitle("Queue")
                        .setColor(0x9F90CF)
                        .setTimestamp(Instant.now());
                if (trackSize == 0) {
                    embed.setDescription("The queue is empty.");
                } else {
                    embed.setDescription(description);
                }
                event.getHook().sendMessageEmbeds(embed.build()).queue();
            }
            case "clear" -> {
                queue.clear();
                MessageEmbed embed = new EmbedBuilder()
                        .setTitle("Queue")
                        .setDescription("Queue cleared")
                        .setColor(0x9F90CF)
                        .setTimestamp(Instant.now())
                        .build();
                event.getHook().sendMessageEmbeds(embed).queue();
            }
        }
    }

}

File: src/main/kotlin/xyz/chalky/taboo/commands/music/QueueSlashCommand.kt
package xyz.chalky.taboo.commands.music

import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import dev.minn.jda.ktx.Embed
import dev.minn.jda.ktx.EmbedBuilder
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.dv8tion.jda.api.interactions.commands.build.SubcommandData
import xyz.chalky.taboo.Taboo
import xyz.chalky.taboo.backend.CommandFlag
import xyz.chalky.taboo.backend.SlashCommand
import xyz.chalky.taboo.util._reply
import xyz.chalky.taboo.util.onSubCommand
import java.time.Instant

class QueueSlashCommand : SlashCommand() {

    init {
        setCommandData(
            Commands.slash("queue", "Queue a song")
                .addSubcommands(
                    SubcommandData(
                        "list", "List the current queue."
                    ),
                    SubcommandData(
                        "clear", "Clear the current queue."
                    )
                )
        )
        addCommandFlags(CommandFlag.MUSIC)
    }

    override fun executeCommand(event: SlashCommandInteractionEvent) {
        val guildAudioPlayer = Taboo.getInstance().audioManager.getAudioPlayer(event.guild!!.idLong)
        val queue = guildAudioPlayer.scheduler.queue
        event.onSubCommand("list") {
            val description = StringBuilder("Queue:\n")
            val trackSize = queue.size
            val trackCount = trackSize.coerceAtMost(10)
            val trackList: List<AudioTrack> = queue.toList()
            for (i in 0 until trackCount) {
                val track = trackList[i]
                val info = track.info
                description.append("`#")
                    .append(i + 1)
                    .append("` ")
                    .append(info.title)
                    .append(" [")
                    .append(info.author)
                    .append("]\n")
            }
            if (trackSize > trackCount) {
                description.append("And ")
                    .append("`")
                    .append(trackSize - trackCount)
                    .append("`")
                    .append(" more tracks...")
            }
            val embed = EmbedBuilder {
                title = "Queue"
                color = 0x9F90CF
                timestamp = Instant.now()
            }
            if (trackSize == 0) {
                embed.description = "The queue is empty."
            } else {
                embed.description = description.toString()
            }
            event._reply(embed.build()).queue()
        }
        event.onSubCommand("clear") {
            queue.clear()
            event._reply(
                Embed {
                    title = "Queue"
                    description = "Queue cleared."
                    color = 0x9F90CF
                    timestamp = Instant.now()
                }
            ).queue()
        }
    }

}